mycroft/skills/skill_loader.py

```diff
@@ -0,0 +1,235 @@
+        self._load()
+
+    def _unload(self):
+        """Remove listeners and stop threads before loading"""
+        self._execute_instance_shutdown()
+        self._garbage_collect()
+        self.loaded = False
+        self._emit_skill_shutdown_event()
+
+    def _execute_instance_shutdown(self):
+        """Call the shutdown method of the skill being reloaded."""
+        try:
+            self.instance.default_shutdown()
+        except Exception as e:
+            log_msg = 'An error occurred while shutting down {}'
+            LOG.error(log_msg.format(self.instance.name))
+            LOG.exception(e)
+        else:
+            LOG.info('Skill {} shut down successfully'.format(self.skill_id))
+
+    def _garbage_collect(self):
+        """Invoke Python garbage collector to remove false references"""
+        if self.config.get("debug", False):
+            gc.collect()
+            # Remove two local references that are known
+            refs = sys.getrefcount(self.instance) - 2
+            if refs > 0:
+                log_msg = (
+                    "After shutdown of {} there are still {} references "
+                    "remaining. The skill won't be cleaned from memory."
+                )
+                LOG.warning(log_msg.format(self.instance.name, refs))
+
+    def _emit_skill_shutdown_event(self):
+        message = Message(
+            "mycroft.skills.shutdown",
+            data=dict(path=self.skill_directory, id=self.skill_id)
+        )
+        self.bus.emit(message)
+
+    def _load(self):
+        self._prepare_for_load()
+        if self.is_blacklisted:
+            self._skip_load()
+        else:
+            skill_module = self._load_skill_source()
+            if skill_module is not None:
+                self._create_skill_instance(skill_module)
+                self._check_for_first_run()
+        self._communicate_load_status()
+
+    def _prepare_for_load(self):
+        self.load_attempted = True
+        self.loaded = False
+        self.instance = None
+
+    def _skip_load(self):
+        log_msg = 'Skill {} is blacklisted - it will not be loaded'
+        LOG.info(log_msg.format(self.skill_id))
+
+    def _load_skill_source(self):
+        """Use Python's import library to load a skill's source code."""
+        # TODO: Replace the deprecated "imp" library with the newer "importlib"
+        module_name = self.skill_id.replace('.', '_')
+        main_file_path = os.path.join(self.skill_directory, SKILL_MAIN_MODULE)
+        try:
+            with open(main_file_path, 'rb') as main_file:
+                skill_module = imp.load_module(
+                    module_name,
+                    main_file,
+                    main_file_path,
+                    ('.py', 'rb', imp.PY_SOURCE)
+                )
+        except FileNotFoundError as f:
+            error_msg = 'Failed to load {} due to a missing file.'
+            LOG.error(error_msg.format(self.skill_id))
```

Comment:
I was confused about how a lot of the log.exception commands pass the exception string when that is not needed.  read up on log.exception and it will print the traceback along with your message and you don't need to pass the exception object.  will fix all of these as I did this all over the refactor